File: soal_1.py
'''
Soal Ujian 1: Aplikasi Pengelolaan Data Produk
Deskripsi Soal:
Buat aplikasi untuk mengelola data produk di toko. Aplikasi ini harus dapat menambahkan produk, menampilkan produk, mencari produk termahal, 
menghitung rata-rata harga, dan mengupdate stok produk.

Pertanyaan:
a. Buat program untuk menambahkan produk baru (nama, harga, stok) ke dalam array/list, dan tampilkan seluruh data setelah produk ditambahkan.
b. Buat program untuk menampilkan semua produk yang ada di toko.
c. Buat program untuk mencari dan menampilkan produk dengan harga tertinggi.
d. Buat program untuk menghitung dan menampilkan rata-rata harga produk.
e. Buat program untuk memperbarui stok produk yang ada di toko berdasarkan nama produk yang dimasukkan pengguna.
'''
# buat list kosong nama product
list_produk = []

# buat fungsi menambah product (soal a)
def tambah_produk():
    nama = input("Masukan nama barang: ")
    harga = int(input("Masukan harga barang: "))
    stok = int(input("Masukan stock barang: "))
    list_produk.append({"nama": nama, "harga": harga, "stok":stok})
    print("Produk Berhasil Ditambahkan")

# buat fungsi tampil product (soal b)
def tampil_produk():    
    for nama_produk in list_produk:
        print(f"Nama item: {nama_produk['nama']}")
        print(f"Harga item: {nama_produk['harga']}")
        print(f"Jumlah stock: {nama_produk['stok']}")

# buat program dengan harga tertinggi (soal c)
def produk_termahal():
    if not list_produk:
        print("Tidak ada produk pada list produk")
        return
    termahal = list_produk[0]
    for nama_produk in list_produk:
        if nama_produk['harga'] > termahal['harga']:
            termahal = nama_produk
    print(f"Produk termahal adalah: Nama: {termahal['nama']}, Harga: {termahal['harga']}, dan Stok {termahal['stok']}")

# buat program untuk menghitung rata-rata harga produk (soal d)
def rata_rata_harga():
    if not list_produk:
        print("Tidak ada produk pada list produk")
        return
    total_harga = 0
    for nama_produk in list_produk:
        total_harga += nama_produk['harga']
    rata2 = total_harga/len(list_produk)
    print(f"Rata-rata harga adalah: {rata2}") 

# buat program untuk update produk (soal e)
def update_stok(nama, stok_update):
    for nama_produk in list_produk:
        if nama_produk['nama'] == nama:
            nama_produk['stok'] = stok_update
            print(f"Nama Produk : {nama} telah diupdate dengan")
            print(f"Stock Baru: {stok_update}")
        else:
            print("Produk tidak ditemukan")

# main program
if __name__ == "__main__":
    jumlah_produk = int(input("Berapa Item yang ingin dimasukan: "))
    print("\n-------soal a-------\n")
    for i in range(jumlah_produk):
        print(f"Produk ke-{i+1} adalah:")
        tambah_produk()
    print("\n-------soal b-------\n")
    tampil_produk()
    print("\n-------soal c-------\n")
    produk_termahal()
    print("\n-------soal d-------\n")
    rata_rata_harga()
    print("\n-------soal e-------\n")
    update_stok("motor", 1)
    tampil_produk()

File: soal_1.c
#include <stdio.h>
#include <string.h>

#define MAX_PRODUK 100

// Struktur data untuk menyimpan produk
struct Produk {
    char nama[50];
    double harga;
    int stok;
};

// Array untuk menyimpan produk
struct Produk list_produk[MAX_PRODUK];
int jumlah_produk = 0;

// Fungsi untuk menambahkan produk baru (soal a)
void tambah_produk() {
    if (jumlah_produk < MAX_PRODUK) {
        printf("Masukkan nama barang: ");
        scanf("%s", list_produk[jumlah_produk].nama);
        printf("Masukkan harga barang: ");
        scanf("%lf", &list_produk[jumlah_produk].harga);
        printf("Masukkan stok barang: ");
        scanf("%d", &list_produk[jumlah_produk].stok);
        jumlah_produk++;
        printf("Produk berhasil ditambahkan.\n");
    } else {
        printf("Kapasitas produk penuh.\n");
    }
}

// Fungsi untuk menampilkan seluruh produk (soal b)
void tampil_produk() {
    printf("\nDaftar produk:\n");
    for (int i = 0; i < jumlah_produk; i++) {
        printf("Nama: %s, Harga: %.2f, Stok: %d\n", list_produk[i].nama, list_produk[i].harga, list_produk[i].stok);
    }
}

// Fungsi untuk mencari dan menampilkan produk termahal (soal c)
void produk_termahal() {
    if (jumlah_produk == 0) {
        printf("Tidak ada produk yang dicatat.\n");
        return;
    }

    struct Produk termahal = list_produk[0];
    for (int i = 1; i < jumlah_produk; i++) {
        if (list_produk[i].harga > termahal.harga) {
            termahal = list_produk[i];
        }
    }

    printf("\nProduk termahal adalah %s dengan harga %.2f dan stok %d\n", termahal.nama, termahal.harga, termahal.stok);
}

// Fungsi untuk menghitung dan menampilkan rata-rata harga produk (soal d)
void rata_rata_harga() {
    if (jumlah_produk == 0) {
        printf("Tidak ada produk yang dicatat.\n");
        return;
    }

    double total_harga = 0;
    for (int i = 0; i < jumlah_produk; i++) {
        total_harga += list_produk[i].harga;
    }

    double rata_rata = total_harga / jumlah_produk;
    printf("\nRata-rata harga produk adalah: %.2f\n", rata_rata);
}

// Fungsi untuk memperbarui stok produk berdasarkan nama produk (soal e)
void update_stok(char nama[], int stok_baru) {
    int found = 0;
    for (int i = 0; i < jumlah_produk; i++) {
        if (strcmp(list_produk[i].nama, nama) == 0) {
            list_produk[i].stok = stok_baru;
            printf("Stok produk %s telah diperbarui menjadi %d\n", nama, stok_baru);
            found = 1;
            break;
        }
    }
    if (!found) {
        printf("Produk dengan nama %s tidak ditemukan.\n", nama);
    }
}

// Main program
int main() {
    int jumlah;
    printf("Berapa banyak produk yang ingin ditambahkan? ");
    scanf("%d", &jumlah);

    printf("\n-------Soal a-------\n");
    for (int i = 0; i < jumlah; i++) {
        printf("Produk ke-%d:\n", i + 1);
        tambah_produk();
    }

    printf("\n-------Soal b-------\n");
    tampil_produk();

    printf("\n-------Soal c-------\n");
    produk_termahal();

    printf("\n-------Soal d-------\n");
    rata_rata_harga();

    printf("\n-------Soal e-------\n");
    char nama_update[50];
    int stok_baru;
    printf("Masukkan nama produk yang ingin diupdate stoknya: ");
    scanf("%s", nama_update);
    printf("Masukkan stok baru: ");
    scanf("%d", &stok_baru);
    update_stok(nama_update, stok_baru);

    printf("\nProduk setelah update stok:\n");
    tampil_produk();

    return 0;
}
